src-ghc/Pact/Types/Crypto.hs

```diff
@@ -217,61 +183,87 @@ instance ConvertBS Ed25519.Signature where
   fromBS = Right . Ed25519.Sig
 #endif
 
-
-
---------- SCHEME ETH INSTANCES --------
-
-instance Scheme (SPPKScheme 'ETH) where
-  type PublicKey (SPPKScheme 'ETH) = ECDSA.PublicKey
-  type PrivateKey (SPPKScheme 'ETH) = ECDSA.PrivateKey
-  type Signature (SPPKScheme 'ETH) = ECDSA.Signature
-
-
-  _sign _ (Hash msg) pub priv = ECDSA.signETH (fromShort msg) pub priv
-  _valid _ (Hash msg) pub sig = ECDSA.validETH (fromShort msg) pub sig
-  _genKeyPair _ = ECDSA.genKeyPair
-  _getPublic _ = Just . ECDSA.getPublicKey
-  _formatPublicKey _ p = ECDSA.formatPublicKeyETH (toBS p)
-
-
-
-instance ConvertBS (ECDSA.PublicKey) where
-  toBS = ECDSA.exportPublic
-  fromBS s = maybeToEither ("Invalid ECDSA Public Key: " ++ show (toB16Text s))
-             (ECDSA.importPublic s)
-instance ConvertBS (ECDSA.PrivateKey) where
-  toBS = ECDSA.exportPrivate
-  fromBS s = maybeToEither ("Invalid ECDSA Private Key: " ++ show (toB16Text s))
-             (ECDSA.importPrivate s)
-instance ConvertBS (ECDSA.Signature) where
-  toBS = ECDSA.exportSignature
-  fromBS s = maybeToEither ("Invalid ECDSA Signature: " ++ show (toB16Text s))
-             (ECDSA.importSignature s)
-
-
-
+instance Scheme (SPPKScheme 'WebAuthn) where
+
+  _valid _ msg (PubBS pubBS) (SigBS sigBS) =
+  -- _valid _ msg UserSig{..} Signer{..} =
+    case runVerification of
+      Left e -> error e
+      Right () -> True
+    where
+      -- Verifying a WebAuthn signature requires that we know the payload
+      -- signed by the WebAuthn keys on the client device. This payload is
+      -- a combination of the Challenge (in our case, the PactHash of a
+      -- transaction), a JSON object of "client metadata", and data about
+      -- the authenticator hardware.
+      -- We require this data to be part of our WebAuthn signature so that
+      -- we can reconstitute the payload that was signed on the browser.
+      runVerification :: Either String ()
+      runVerification = do
+
+        -- Decode our WebAuthn signature object from a `UserSig` string.
+        WebAuthnSignature
+          { authenticatorData
+          , signature
+          , clientDataJSON } <- A.eitherDecode (BSL.fromStrict sigBS)
+        -- Decode the signer's public key.
+        publicKeyBytes <- return pubBS -- Base16.decode (T.encodeUtf8 pubBS)
+        publicKey <- first show $ Serialise.deserialiseOrFail @WA.CosePublicKey (BSL.fromStrict publicKeyBytes)
```

Comment:
Usage of `show` feels strange here

Reply:
The `show` is only for the error case. `first` here is doing the job of `mapError`. It's turning a `DeserializeError` into a `String`.

In the commit you are reviewing, that's a fairly big issue because ultimately we call `error` on that string, which itself is a mistake.

So I will push a commit that fixes that issue: if validation fails, we will return `False` instead of panicking :)

Reply:
After the fix, do you think we should still remove the `show`?